lib/plugins/aws/lib/naming.js

```diff
@@ -0,0 +1,234 @@
+ *
+ * There are two main concepts:
+ * 1. normalization - normalization of a name comprises removing any unacceptable characters and
+ * changing the name to comply with capitalization expectations.
+ * 2. logicalization - logicalization of a name comprises preparing it for use as an attribute name
+ * for a resource within the Service's CloudFormation Template.  Notably, a suffix or prefix should
+ * be added to the name of resources.  These help to disambiguate attributes identifying or
+ * declaring different aspects or data about the same resource.  (e.g. MyFuncLambdaFunction vs.
+ * MyFuncLambdaFunctionArn)  Generally, normalization of the name also occurs.
+ *
+ * Sometimes, it is important to deconstruct the names that are created via this centralized
+ * utility.  As a result, the regular expressions used for identifying specific resource types using
+ * their names or the logic to extract a name components from a logical name are things that must
+ * also be centralized so that they co-occur and are more correlated spatially within the code base.
+ * That is to say that they are easier to change together.
+ */
+class Naming {
+  // TODO this configure pattern is fugly.  it serves the need to have static and also configured
+  // TODO use patterns without complicating the life of consumers.  In standard (non-test) usage
+  // TODO the frequency of using this method should be low so maybe it doesn't matter.
+  // TODO better suggestions highly welcome
+  configure(serverless) {
+    this.serverless = serverless;
+  }
+
+  /*
+   * General Name Normalization
+   */
+  normalizeName(name) {
+    return `${_.upperFirst(name)}`;
+  }
+  normalizeNameToAlphaNumericOnly(name) {
+    return this.normalizeName(name.replace(/[^0-9A-Za-z]/g, ''));
+  }
+  normalizePathtoCapitalAlphaNumbericOnlyWithReplacement(path) {
+    return this.normalizeNameToAlphaNumericOnly(
+      path.replace(/-/g, 'Dash')
+          .replace(/\{(.*)\}/g, '$1Var'));
+  }
+
+  // Probably belongs in a non-aws-specific code asset
+  getServiceEndpointRegex() {
+    return /^ServiceEndpoint/;
+  }
+
+  /*
+   * Stack Naming
+   */
+  getStackName() {
+    return `${this.serverless.service.service}-${this.serverless.config.stage}`;
+  }
+
+  /*
+   * IAM (Role and Policy) Naming
+   */
+  getRolePath() {
+    return '/';
+  }
+  getRoleName() {
+    return {
+      'Fn::Join': [
+        '-',
+        [
+          this.serverless.service.service,
+          this.serverless.config.stage,
+          this.serverless.config.region,
+          'lambdaRole',
+        ],
+      ],
+    };
+  }
+  getLogicalRoleName() {
+    return 'IamRoleLambdaExecution';
+  }
+  getPolicyName() {
+    return { // TODO should probably have name ordered and altered as above - see AWS docs
+      'Fn::Join': [
+        '-',
+        [
+          this.serverless.config.stage,
+          // TODO is it a bug here in that the role is not specific to the region?
```

Comment:
@erikerikson that's a super interesting user case! I haven't thought about that before! Do you think this would cause an issue for you guys at Nordstrom?